src/server/game/Anticheat/AnticheatMgr.h

```diff
@@ -0,0 +1,151 @@
+/*
+* Copyright (C) 2011-2016 TrinityCore <http://www.trinitycore.org/>
+*
+* This program is free software; you can redistribute it and/or modify it
+* under the terms of the GNU General Public License as published by the
+* Free Software Foundation; either version 2 of the License, or (at your
+* option) any later version.
+*
+* This program is distributed in the hope that it will be useful, but WITHOUT
+* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+* FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+* more details.
+*
+* You should have received a copy of the GNU General Public License along
+* with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef SC_ACMGR_H
+#define SC_ACMGR_H
+
+#include "AnticheatData.h"
+#include "Common.h"
+#include "SharedDefines.h"
+#include "ScriptMgr.h"
+#include "Player.h"
+
+class ChatHandler;
+class AnticheatCheck;
+
+enum DetectionTypes
+{
+    SPEED_HACK_DETECTION            = 0x01,
+    FLY_HACK_DETECTION              = 0x02,
+    WALK_WATER_HACK_DETECTION       = 0x04,
+    JUMP_HACK_DETECTION             = 0x08,
+    TELEPORT_PLANE_HACK_DETECTION   = 0x10,
+    CLIMB_HACK_DETECTION            = 0x20
+};
+
+typedef std::unordered_map<ObjectGuid, AnticheatData> AnticheatPlayersDataMap;
+
+class AnticheatMgr
+{
+    friend class AnticheatCheck;
+
+    AnticheatMgr();
+    ~AnticheatMgr();
+
+    public:
+        static AnticheatMgr* instance()
+        {
+           static AnticheatMgr instance;
+           return &instance;
+        }
+
+        void StartHackDetection(Player* player, MovementInfo const& movementInfo, uint16 opcode);
+
+        void HandlePlayerLogin(Player* player, PreparedQueryResult result);
+        void HandlePlayerLogout(ObjectGuid const& guid);
+        void SavePlayerData(ObjectGuid const& guid, SQLTransaction& trans);
+
+        inline uint32 GetTotalReports(ObjectGuid const& guid) { return m_playerMap[guid].GetTotalReports(); }
+        inline float GetAverage(ObjectGuid const& guid) { return m_playerMap[guid].GetAverage(); }
+        inline uint32 GetTypeReports(ObjectGuid const& guid, ReportTypes type) { return m_playerMap[guid].GetTypeReports(type); }
+
+        // use sparingly, triggers a global anticheat save to DB, requires to lock player map and does some synchronous queries
+        bool AnticheatGlobalCommand(ChatHandler* handler);
+
+        void ResetDailyReportStates();
+
+        void LoadSettings();
+
+    private:
+        AnticheatData* GetPlayerData(ObjectGuid const& guid)
+        {
+            auto itr = m_playerMap.find(guid);
+            if (itr != m_playerMap.end())
+                return &itr->second;
+            return nullptr;
+        }
+
+        AnticheatPlayersDataMap m_playerMap;
```

Comment:
I always wonder if we should keep maps using Player guid as key or just store the data in Player class itself